File: engine/packages/capacitor-modoki-system/android/src/main/java/com/modokiengine/capacitor/system/ModokiSystemPlugin.java
package com.modokiengine.capacitor.system;

import android.content.ActivityNotFoundException;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.provider.Settings;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.google.android.play.core.review.ReviewInfo;
import com.google.android.play.core.review.ReviewManager;
import com.google.android.play.core.review.ReviewManagerFactory;

@CapacitorPlugin(name = "ModokiSystem")
public class ModokiSystemPlugin extends Plugin {

    @PluginMethod
    public void openAppSettings(PluginCall call) {
        String target = call.getString("target", "app");
        Context context = getContext();
        String packageName = context.getPackageName();

        // An app-settings URI cannot express either page: both are ACTION intents, which is why
        // this is native at all rather than a link the WebView hands to the system.
        Intent details = new Intent(
            Settings.ACTION_APPLICATION_DETAILS_SETTINGS, Uri.fromParts("package", packageName, null));
        Intent intent = details;
        if ("notifications".equals(target) && Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            intent = new Intent(Settings.ACTION_APP_NOTIFICATION_SETTINGS);
            intent.putExtra(Settings.EXTRA_APP_PACKAGE, packageName);
        }

        JSObject ret = new JSObject();
        ret.put("opened", start(intent) || (intent != details && start(details)));
        call.resolve(ret);
    }

    // A web page in the default browser (#1196). https only: a VIEW intent would also resolve
    // `intent:`, `tel:` or another app's custom scheme, which an authored link must never reach.
    // No <queries> entry is needed — startActivity is called directly and a device with no
    // browser surfaces as ActivityNotFoundException, which start() turns into false.
    @PluginMethod
    public void openUrl(PluginCall call) {
        String raw = call.getString("url");
        Uri uri = raw == null ? null : Uri.parse(raw);
        boolean openable = uri != null
            && "https".equalsIgnoreCase(uri.getScheme())
            && uri.getHost() != null
            && !uri.getHost().isEmpty();

        JSObject ret = new JSObject();
        ret.put("opened", openable && start(new Intent(Intent.ACTION_VIEW, uri)));
        call.resolve(ret);
    }

    // Plain text on the system clipboard (#1398) — a player ID the player pastes into a support
    // email. Native because a WebView clipboard write needs the tap's user activation, which an
    // engine-dispatched action cannot promise. Android 13+ shows its own "Copied" confirmation.
    @PluginMethod
    public void copyText(PluginCall call) {
        String text = call.getString("text");
        JSObject ret = new JSObject();
        ClipboardManager clipboard = (ClipboardManager) getContext().getSystemService(Context.CLIPBOARD_SERVICE);
        if (text == null || text.isEmpty() || clipboard == null) {
            ret.put("copied", false);
            call.resolve(ret);
            return;
        }
        try {
            clipboard.setPrimaryClip(ClipData.newPlainText("text", text));
            ret.put("copied", true);
        } catch (RuntimeException e) {
            // A background app or a restricted profile can refuse the write.
            ret.put("copied", false);
        }
        call.resolve(ret);
    }

    // The Play In-App Review flow (#939). Two async steps: request the ReviewInfo, then launch it.
    // `requested` means only that Play accepted the launch — it never reports whether a dialog was
    // shown or a review written, and it silently does nothing once its own quota is spent. Any
    // failure (no Play services, a sideloaded build, an internal Play error) resolves false rather
    // than rejecting, so the caller needs no platform branch.
    @PluginMethod
    public void requestReview(PluginCall call) {
        final android.app.Activity activity = getActivity();
        if (activity == null) {
            resolveRequested(call, false);
            return;
        }
        try {
            final ReviewManager manager = ReviewManagerFactory.create(getContext());
            manager.requestReviewFlow().addOnCompleteListener(task -> {
                if (!task.isSuccessful()) {
                    resolveRequested(call, false);
                    return;
                }
                ReviewInfo info = task.getResult();
                manager.launchReviewFlow(activity, info)
                    .addOnCompleteListener(flow -> resolveRequested(call, flow.isSuccessful()));
            });
        } catch (Exception e) {
            resolveRequested(call, false);
        }
    }

    private void resolveRequested(PluginCall call, boolean requested) {
        JSObject ret = new JSObject();
        ret.put("requested", requested);
        call.resolve(ret);
    }

    private boolean start(Intent intent) {
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        try {
            getContext().startActivity(intent);
            return true;
        } catch (ActivityNotFoundException e) {
            return false;
        }
    }

    // The backup-excluded key-value store is iOS-only (#1271): on Android the whole app's backup is
    // off (#1267), so PlayerPrefs stays on SharedPreferences and the engine never calls these. They
    // exist so the plugin's JS contract dispatches somewhere on every platform, and say why they fail.
    private static final String KV_IOS_ONLY =
        "The backup-excluded store is iOS-only; Android keeps PlayerPrefs in SharedPreferences with backup off.";

    @PluginMethod
    public void kvGetAll(PluginCall call) {
        call.unavailable(KV_IOS_ONLY);
    }

    @PluginMethod
    public void kvSet(PluginCall call) {
        call.unavailable(KV_IOS_ONLY);
    }

    @PluginMethod
    public void kvRemove(PluginCall call) {
        call.unavailable(KV_IOS_ONLY);
    }

    @PluginMethod
    public void kvInfo(PluginCall call) {
        call.unavailable(KV_IOS_ONLY);
    }
}

File: engine/packages/capacitor-modoki-system/ios/Sources/ModokiSystemPlugin/ModokiSystemPlugin.swift
import Capacitor
import StoreKit
import UIKit

@objc(ModokiSystemPlugin)
public class ModokiSystemPlugin: CAPPlugin, CAPBridgedPlugin {

    public let identifier = "ModokiSystemPlugin"
    public let jsName = "ModokiSystem"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "openAppSettings", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openUrl", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "copyText", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "kvGetAll", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "kvSet", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "kvRemove", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "kvInfo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestReview", returnType: CAPPluginReturnPromise),
    ]

    private let store = BackupExcludedStore()

    // The backup-excluded key-value store PlayerPrefs uses on iOS (#1271) — see BackupExcludedStore.
    @objc func kvGetAll(_ call: CAPPluginCall) {
        do {
            call.resolve(["entries": try store.getAll(prefix: call.getString("prefix") ?? "")])
        } catch {
            call.reject("kvGetAll failed: \(error.localizedDescription)")
        }
    }

    @objc func kvSet(_ call: CAPPluginCall) {
        guard let key = call.getString("key"), let value = call.getString("value") else {
            call.reject("kvSet needs a string key and value")
            return
        }
        do {
            try store.set(key: key, value: value)
            call.resolve()
        } catch {
            call.reject("kvSet failed: \(error.localizedDescription)")
        }
    }

    @objc func kvRemove(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            call.reject("kvRemove needs a string key")
            return
        }
        do {
            try store.remove(key: key)
            call.resolve()
        } catch {
            call.reject("kvRemove failed: \(error.localizedDescription)")
        }
    }

    @objc func kvInfo(_ call: CAPPluginCall) {
        do {
            call.resolve(try store.info())
        } catch {
            call.reject("kvInfo failed: \(error.localizedDescription)")
        }
    }

    // Plain text on the general pasteboard (#1398) — a player ID the player pastes into a support
    // email. Native because a web-view clipboard write needs the tap's user activation, which an
    // engine-dispatched action cannot promise. UIPasteboard is main-thread UIKit.
    @objc func copyText(_ call: CAPPluginCall) {
        guard let text = call.getString("text"), !text.isEmpty else {
            call.resolve(["copied": false])
            return
        }
        DispatchQueue.main.async {
            UIPasteboard.general.string = text
            call.resolve(["copied": UIPasteboard.general.hasStrings])
        }
    }

    // A web page in Safari (#1196). https only: `UIApplication.open` would also act on `tel:`,
    // `sms:` or another app's custom scheme, which an authored link must never reach.
    @objc func openUrl(_ call: CAPPluginCall) {
        guard let raw = call.getString("url"),
              let url = URL(string: raw),
              url.scheme?.lowercased() == "https",
              url.host != nil else {
            call.resolve(["opened": false])
            return
        }
        DispatchQueue.main.async {
            UIApplication.shared.open(url, options: [:]) { opened in
                call.resolve(["opened": opened])
            }
        }
    }

    @objc func openAppSettings(_ call: CAPPluginCall) {
        let target = call.getString("target") ?? "app"
        DispatchQueue.main.async {
            var urlString = UIApplication.openSettingsURLString
            // The notification-specific page exists only from iOS 16; older versions land on the
            // app's own Settings page, which lists Notifications one tap further in.
            if target == "notifications", #available(iOS 16.0, *) {
                urlString = UIApplication.openNotificationSettingsURLString
            }
            guard let url = URL(string: urlString) else {
                call.resolve(["opened": false])
                return
            }
            UIApplication.shared.open(url, options: [:]) { opened in
                call.resolve(["opened": opened])
            }
        }
    }

    // The OS review prompt (#939). `requested` says only that the request was handed over: StoreKit
    // never reports whether the sheet appeared, and it silently does nothing once its own quota is
    // spent. Resolving false without a window scene rather than rejecting keeps the caller free of
    // platform branching.
    @objc func requestReview(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            // compactMap BEFORE the activation test: `first(where:)` then `as?` takes the first
            // foreground-active scene of ANY type and resolves false if it is not a window scene,
            // although a valid one exists. The ask is already stamped as spent by then, so that
            // costs the install its one opportunity and reports nothing.
            guard let scene = UIApplication.shared.connectedScenes
                .compactMap({ $0 as? UIWindowScene })
                .first(where: { $0.activationState == .foregroundActive }) else {
                call.resolve(["requested": false])
                return
            }
            SKStoreReviewController.requestReview(in: scene)
            call.resolve(["requested": true])
        }
    }
}
